### PR TITLE
Don't remove unused selection if referenced in GROUP BY

### PR DESCRIPTION
**Pull request by barakalon**

### CODE REVIEW

File: tests/fixtures/optimizer/tpc-h/tpc-h.sql [hunk 1]
```diff
@@ -1043,6 +1043,7 @@ SELECT
   COUNT(*) AS "custdist"
 FROM (
   SELECT
+    "customer"."c_custkey" AS "c_custkey",
```
Comment (barakalon) [inline tests/fixtures/optimizer/tpc-h/tpc-h.sql:1046]: This isn't really ideal. But right now we support queries like this:

```sql
SELECT 
  x.b, 
  x.a 
FROM x 
LEFT JOIN y 
GROUP BY b
```

"b" is ambiguous, but some dialects might assume "b" is the output column "x.b". At least we have tests that make that assumption...

File: sqlglot/optimizer/qualify_columns.py [hunk 2]
```diff
@@ -122,6 +123,30 @@ def _expand_using(scope, schema):
...
 
+def _expand_group_numbers(scope):
+    selects = scope.expression.args.get("expressions")
+    group = scope.expression.args.get("group")
+    if not group:
+        return
+
+    new_expressions = []
+    for expression in group.args.get("expressions"):
+        if isinstance(expression, exp.Literal) and expression.is_int:
```
Comment (tobymao) [inline sqlglot/optimizer/qualify_columns.py:134]: do you mind refactoring so is_int and is_string is on base object so we don't always have to check if the expression is a literal?

File: sqlglot/optimizer/qualify_columns.py [hunk 3]
```diff
@@ -122,6 +123,30 @@ def _expand_using(scope, schema):
...
+            try:
+                select = selects[int(expression.this) - 1]
+            except IndexError:
+                # pylint: disable=raise-missing-from
+                raise OptimizeError(f"Unknown output column: {expression.name}")
+            if isinstance(select, exp.Alias):
+                select = select.this
+            new_expressions.append(select.copy())
```
Comment (tobymao) [inline sqlglot/optimizer/qualify_columns.py:142]: what's this copy for?

File: sqlglot/optimizer/qualify_columns.py [hunk 4]
```diff
@@ -122,6 +123,30 @@ def _expand_using(scope, schema):
                 scope.replace(column, replacement)
 
 
+def _expand_group_numbers(scope):
```
Comment (tobymao) [inline sqlglot/optimizer/qualify_columns.py:126]: do we want to expand group numbers?

File: sqlglot/optimizer/qualify_columns.py (hunk 3 above)
Comment (barakalon) [reply sqlglot/optimizer/qualify_columns.py:142]: we need to copy the expression, or else we'll have one object with 2 parents (well, 1 parent, but I think you get the point)

**Comment by barakalon** [comment]
Alright, this should be ready for another look

File: sqlglot/optimizer/qualify_columns.py [hunk 5]
```diff
@@ -122,40 +125,87 @@ def _expand_using(scope, schema):
...
+            select = selects.get(node.name)
+            if select:
+                scope.clear_cache()
+                if isinstance(select, exp.Alias):
+                    select = select.this
+                return select.copy()
+
+        return node
+
+    group = group.transform(transform)
```
Comment (tobymao) [inline sqlglot/optimizer/qualify_columns.py:154]: transform is pretty expensive, can we do a copy=False here?

File: sqlglot/optimizer/qualify_columns.py [hunk 6]
```diff
@@ -122,40 +125,87 @@ def _expand_using(scope, schema):
...
+    ):
+        node.replace(new_node)
+
+
+def _expand_positional_references(scope, expressions):
+    new_nodes = []
+    for node in expressions:
+        if node.is_int:
+            try:
+                select = scope.selects[int(node.this) - 1]
```
Comment (tobymao) [inline sqlglot/optimizer/qualify_columns.py:177]: node.name since it's for sure the text representation

**Review by tobymao — APPROVED** [review]
can you do a quick perf check on qualify columns to see that perf hasn't completely tanked?

File: sqlglot/optimizer/qualify_columns.py (hunk 5 above)
Comment (tobymao) [inline sqlglot/optimizer/qualify_columns.py:154]: should we do a copy=False here for perf?

File: sqlglot/optimizer/qualify_columns.py [hunk 7]
```diff
@@ -122,40 +125,87 @@ def _expand_using(scope, schema):
...
+
 
+def _expand_order_by(scope):
+    order = scope.expression.args.get("order")
+    if not order:
+        return
+
+    expressions = [o.this for o in order.expressions]
+    for node, new_node in zip(
+        expressions, _expand_positional_references(scope, expressions)
```
Comment (tobymao) [inline sqlglot/optimizer/qualify_columns.py:167]: expressions can just be an inline generator (o.this ...)

Comment (tobymao) [reply sqlglot/optimizer/qualify_columns.py:167]: or just do zip(order.expressions) and then 169 ordered.this.replace

Comment (barakalon) [reply sqlglot/optimizer/qualify_columns.py]: I'm not sure I follow completely, let me know if the replacement is OK.

**Comment by barakalon** [comment]
I'm not seeing a noticeable performance hit.

Here's what I'm using to benchmark:

```python
from sqlglot import parse_one
from sqlglot.optimizer import optimize
import timeit


from tests.helpers import TPCH_SCHEMA, load_sql_fixture_pairs


def optimize_tpc_h():
    for _, sql, _ in load_sql_fixture_pairs(f"optimizer/tpc-h/tpc-h.sql"):
        optimize(
            expression=parse_one(sql),
            schema=TPCH_SCHEMA
        )


print(timeit.timeit(optimize_tpc_h, number=100))
```

File: sqlglot/optimizer/qualify_columns.py (hunk 7 above)
Comment (tobymao) [reply sqlglot/optimizer/qualify_columns.py]: i didn't realize you use it twice, what you had was fine, sorry for the swirl

**Pull request merged**